transport/ice/IceSession.cpp

```diff
@@ -472,11 +493,252 @@ bool IceSession::isResponseAuthentic(const void* data, size_t len) const
         stunMessage->isAuthentic(_hmacComputer.remote);
 }
 
+bool IceSession::isIceAuthentic(const void* data, size_t len) const
+{
+    if (ice::isResponse(data))
+    {
+        return isResponseAuthentic(data, len);
+    }
+
+    if (ice::isRequest(data))
+    {
+        return isRequestAuthentic(data, len);
+    }
+
+    return false;
+}
+
+bool IceSession::processValidStunTcpRequest(IceEndpoint* localEndpoint,
+    const transport::SocketAddress& remotePort,
+    const StunMessage& msg,
+    int remoteCandidatePriority,
+    uint64_t now)
+{
+    assert(localEndpoint->getTransportType() == TransportType::TCP);
+
+    if (isAttached(localEndpoint))
+    {
+        sendResponse(localEndpoint, remotePort, 0, msg, now);
+        return true;
+    }
+
+    IceCandidate remoteCandidate(_component,
+        localEndpoint->getTransportType(),
+        remoteCandidatePriority,
+        remotePort,
+        remotePort,
+        IceCandidate::Type::PRFLX);
+
+    auto* candidatePair = addRemoteTcpCandidateAndCreatePair(remoteCandidate, localEndpoint);
+    if (candidatePair)
+    {
+        candidatePair->accept();
+        sendResponse(localEndpoint, remotePort, 0, msg, now);
+        candidatePair->receptionTimestamp = now;
+        candidatePair->send(now);
+    }
+
+    if (_state == State::CONNECTING)
+    {
+        sortCheckList();
+    }
+
+    return candidatePair != nullptr;
+}
+
+bool IceSession::processValidStunUdpRequest(IceEndpoint* localEndpoint,
+    const transport::SocketAddress& remotePort,
+    const StunMessage& msg,
+    int remoteCandidatePriority,
+    uint64_t now)
+{
+    assert(localEndpoint->getTransportType() == TransportType::UDP);
+
+    if (!(_state == State::CONNECTING || _state == State::CONNECTED))
+    {
+        return false;
+    }
+
+    auto* pair = findCandidatePair(localEndpoint, remotePort);
+    if (pair)
+    {
+        // Only send response after accept! Otherwise we could drop early media
+        pair->accept();
+        sendResponse(localEndpoint, remotePort, 0, msg, now);
+        return true;
+    }
+
+    const auto* remoteCandidate = findCandidateWithUdpAddress(_remoteCandidates, remotePort);
+    if (remoteCandidate)
+    {
+        // if it is not a new candidate but we don't have a candidate that could mean we are receiving
+        // a request from a pair we have discarded previously.
+        // I don't see why this should happen but if we are receiving a probe from it, perhaps there is no reason to be
+        // discarded. Let's add it again and log a warn
+        const auto* localEndpointInfo = findEndpointInfo(localEndpoint);
+        assert(localEndpointInfo);
+        if (!localEndpointInfo)
+        {
+            // We can't accept endpoint that are not attached. This should be an error as it is not supposed to
+            // happen
+            // This perhaps should send an error instead of just ignore
+            return false;
```

Comment:
Perhaps this should send an error response. It is not supposed to receive requests from a local endpoint that is not attached

Reply:
yes perhaps. It is easier to debug and remote side understands the candidate is undesired. A sendResponse with error should suffice.